docs/theory.rst

```diff
@@ -0,0 +1,248 @@
+
+The annotated test
+++++++++++++++++++
+
+
+And here is the same test, with added trace statements and comments:
+
+.. code-block:: python
+
+    def test_simple_type_inferencing(self):
+        code = dedent("""\
+            class Sample(object):
+                pass
+            a_var = Sample()
+        """)
+
+        trace = True
+
+        def banner(s):
+            if trace:
+                print(f"\n{g._caller_name(2)}: ===== {s}")
+
+
+        banner('after setUp')
+
+        # 1. setUp creates self.project.
+
+            # setUp instantiates self.project to a Project instance.
+            # self.project = testutils.sample_project()
+
+        # 2. get_string_scope sets self.scope to the scope of the test string.
+
+            # Sets self.scope to pyobjectsdef.PyModule(project.pycore, code, ...)
+            # (code is the test string, defined above.)
+
+        # ??? Instantiating the pyobjectsdef.PyModule does all the work ???
+
+            # PyDefinedObject.__init__ calls:
+
+                # self.concluded_attributes = self.get_module()._get_concluded_data()
+                # self.attributes = self.get_module()._get_concluded_data()
+
+            # But all attributes are empty for this test.
+
+        scope = libutils.get_string_scope(self.project, code)
+
+        banner('after get_string_scope\n')
+
+            # scope is a GlobalScope.  It might be any subclass of Scope.
+            # scope.pyobject is a pyobjectsdef.PyModule.
+
+        # if trace: g.trace('*** scope.pyobject', scope.pyobject)
+
+        # *** Calling scope["Sample"] (via _ScopeVisitor._ClassDef)
+        #     instantiates pyobjects.PyClass *and* pyobjectsdef.PyClass.
+        #     (Because pyobjectsDef.PyClass is a subclass of pyobjects.PyClass.)
+        # scope["Sample"] is a pynamesdef.DefinedName.
+
+        sample_class = scope["Sample"].get_object()
+
+        banner('after sample_class = scope["Sample"].get_object()\n')
+
+
+            # sample_class is a pyobjectsdef.PyClass ("::Sample" at ...)
+            # scope["Sample"] is a DefinedName.
+            # scope["Sample"].pyobject is a pyobjectsdef.PyClass.
+
+        a_var = scope["a_var"].get_object()
+
+            # a_var is a pyobjects.PyObject
+            # a_var.get_type() is a pyobjectsdef.PyClass ("::Sample" at ...)
+
+            # scope["a_var"] is an pynamesdef.AssignedName.
+            # scope["a_var"].pyobject is a pynames._Inferred.
+            # scope["a_var"].get_object() is a pyobjects.PyObject.
+
+        if trace:
+            print('')
+            print(f"sample_class: {sample_class}")
+
+        self.assertEqual(sample_class, a_var.get_type())
+
+setUp simulates Rope's startup logic
+++++++++++++++++++++++++++++++++++++
+
+
+`ObjectInferTest.setUp` executes a simplified version of Rope's startup logic:
+
+.. code-block:: python
+
+    def setUp(self):
+        super().setUp()
+        self.project = testutils.sample_project()
```

Comment:
sample_project() really just creates an empty folder and setup a Rope project at a temporary directory, that's all it does. setUp() sometimes would seed additional .py files or python packages if the test needs them.

Note also that there's also an alternative version of this fixture written as pytest-style fixture, rather than unittest-style fixture. Once you get what one does, the other should be fairly straightforward to understand.

It's not really representative of Rope's startup sequence. Actual editor integration would often call one of the project validation methods to additionally ensure that any caches are still valid. Not all operations actually **need** the cache though. Many refactoring operations would just do all the necessary inferences when they needed it.